library/src/main/java/co/fusionx/relay/util/Utils.java

```diff
@@ -8,89 +14,157 @@ public static boolean isNotEmpty(final CharSequence cs) {
+            mFgColor = fg;
+            mBgColor = bg;
+        }
 
-            final int secondDigit = Utils.digitValue(line.charAt(i));
-            if (secondDigit >= 0) {
-                i++;
+        void apply(int pos) {
+            if (pos == 0) {
+                return;
+            }
+            if (mColorStart >= 0 && mColorStart < pos && mFgColor != null) {
+                addFormat(new FormatSpanInfo(mColorStart, pos, mFgColor, mBgColor));
+            }
+            if (mBoldStart >= 0 && mBoldStart < pos) {
+                addFormat(new FormatSpanInfo(mBoldStart, pos, FormatSpanInfo.Format.BOLD));
+            }
+            if (mItalicStart >= 0 && mItalicStart < pos) {
+                addFormat(new FormatSpanInfo(mItalicStart, pos,
+                        FormatSpanInfo.Format.ITALIC));
+            }
+            if (mUnderlineStart >= 0 && mUnderlineStart < pos) {
+                addFormat(new FormatSpanInfo(mUnderlineStart, pos,
+                        FormatSpanInfo.Format.UNDERLINED));
+            }
+        }
 
-                // If out of bounds simply break
-                if (i == length) {
-                    break;
-                }
+        private void addFormat(FormatSpanInfo info) {
+            if (mFormats == null) {
+                mFormats = new ArrayList<>();
             }
+            mFormats.add(info);
+        }
+    }
 
-            // Now skip ",x" or ",xy" (background color).
-            char ch = line.charAt(i);
+    public static Pair<String, List<FormatSpanInfo>> parseAndStripColorsFromMessage(
+            final String line) {
+        final FormatState state = new FormatState();
+        final int length = line.length();
+        final StringBuilder buffer = new StringBuilder();
+        int i = 0;
 
-            // If not a comma simply go to the next one
-            if (ch != ',') {
-                continue;
-            }
 
-            // Advance
-            i++;
+        while (i < length) {
+            char c = line.charAt(i++);
+
+            if (c == '\u0002') {
+                // bold
+                state.bold(buffer.length());
+            } else if (c == '\u0003') {
+                // color
+
+                int firstFgDigit = digitValue(line, i);
+                int secondFgDigit = digitValue(line, i + 1);
+                // advance over read digits
+                if (firstFgDigit >= 0) {
+                    i++;
+                }
+                if (secondFgDigit >= 0) {
+                    i++;
+                }
 
-            if (i == length) {
-                // Keep the comma
-                i--;
-            } else {
-                final int backFirstDigit = Utils.digitValue(line.charAt(i));
-                if (backFirstDigit >= 0) {
+                FormatSpanInfo.Color fg = makeColor(firstFgDigit, secondFgDigit);
+                FormatSpanInfo.Color bg = null;
+                if (i < line.length() && line.charAt(i) == ',') {
```

Comment:
Should check for i < line.length() + 1 here to catch the comma at end of line scenario